File: backend/utils/card_code.py
import hashlib
import secrets

from config import CARD_PEPPER, CARD_PREVIOUS_PEPPERS

CHARSET = 'ABCDEFGHJKMNPQRSTUVWXYZ23456789'

def generate_code() -> str:
    chars = []
    while len(chars) < 16:
        b = secrets.randbits(8)
        if b < 240: chars.append(CHARSET[b % 30])
    s = ''.join(chars)
    return f"{s[0:4]}-{s[4:8]}-{s[8:12]}-{s[12:16]}"

def normalize_code_text(text: str) -> str:
    return text.replace('-', '').upper()

def hash_text(text: str, pepper: str | None = None) -> str:
    return hashlib.sha256(((pepper or CARD_PEPPER) + normalize_code_text(text)).encode()).hexdigest()

def hash_candidates(text: str) -> list[str]:
    hashes = [hash_text(text, CARD_PEPPER)]
    for pepper in CARD_PREVIOUS_PEPPERS:
        digest = hash_text(text, pepper)
        if digest not in hashes:
            hashes.append(digest)
    return hashes

File: packages/backend/src/utils/card-code.ts
import crypto from 'crypto';
import { config } from '../config';

const CHARSET = 'ABCDEFGHJKMNPQRSTUVWXYZ23456789'; // 30 chars, no 0/O/1/I/L

export function generateCardCode(): string {
  const chars: string[] = [];
  while (chars.length < 16) {
    const byte = crypto.randomBytes(1)[0];
    if (byte < 240) { // 240 = 30 * 8, 避免取模偏差
      chars.push(CHARSET[byte % 30]);
    }
  }
  // 格式: XXXX-XXXX-XXXX-XXXX
  return `${chars.slice(0, 4).join('')}-${chars.slice(4, 8).join('')}-${chars.slice(8, 12).join('')}-${chars.slice(12, 16).join('')}`;
}

export function hashCardCode(code: string): string {
  const normalized = code.replace(/-/g, '').toUpperCase();
  return crypto.createHash('sha256').update(config.CARD_PEPPER + normalized).digest('hex');
}

export function normalizeCardCode(code: string): string {
  return code.replace(/-/g, '').toUpperCase();
}
